python/lsst/daf/butler/datastores/inMemoryDatastore.py

```diff
@@ -354,28 +375,47 @@ def remove(self, ref):
         ------
         FileNotFoundError
             Attempt to remove a dataset that does not exist.
+        """
+
+        log.debug("Trash %s in datastore %s", ref, self.name)
+
+        # Check that this dataset is known to datastore
+        self._get_dataset_info(ref)
```

Comment:
Is this check necessary?  It'd probably be safer to make the deletion do nothing if the `ref` isn't actually in the datastore, to guard against the case where the a concurrent delete happens between the check and the move to trash.

Reply:
In this particular case a concurrent delete should be impossible because in memory datastore is constrained to a particular python process. For the other datastores, I don't really know how to respond to multiple processes all trying to delete the same dataset. It sort of means that you have to say that `trash()` is never allowed to fail because you can never really tell if the file should be there or not.